src/main/kotlin/org/jetbrains/research/tasktracker/tracking/activity/ActivityTracker.kt

```diff
@@ -0,0 +1,123 @@
+                if (listenShortcut) {
+                    actionId.extractShortcut(event, id)
+                }
+                activityLogger.log(Type.Action, actionId, id)
+            }
+        }
+        actionListener.connect(AnActionListener.TOPIC)
+    }
+
+    private fun String.extractShortcut(event: AnActionEvent, id: Int) {
+        if (!movingActions.contains(this) && !typingActions.contains(this)) {
+            val input = event.inputEvent
+            if (input is KeyEvent) {
+                activityLogger.log(Type.Shortcut, eventToShortcutInfo(input), id)
+            }
+        }
+    }
+
+    private fun eventToShortcutInfo(event: KeyEvent): String {
+        val keys = Key.values().mapNotNull { it.check(event) }.map { it.name }.plus(event.keyCode.toString())
+        return keys.joinToString(" ")
+    }
+
+    private fun listenExecution() {
+        val executionListener = object : ExecutionListener {
+            override fun processStarting(executorId: String, env: ExecutionEnvironment, handler: ProcessHandler) {
+                val commandLine = (handler as? BaseProcessHandler<*>)?.commandLine ?: ""
+                activityLogger.log(Type.Execution, "$executorId:'${env.runProfile}':$commandLine")
+            }
+        }
+        executionListener.connect(ExecutionManager.EXECUTION_TOPIC)
+    }
+
+    private fun listenKeyboard() {
+        IdeEventQueue.getInstance().addPostprocessor({ awtEvent: AWTEvent ->
+            if (awtEvent is KeyEvent && awtEvent.id == KeyEvent.KEY_PRESSED) {
+                activityLogger.log(Type.KeyPressed, awtEvent.info())
+            }
+            if (awtEvent is KeyEvent && awtEvent.id == KeyEvent.KEY_RELEASED) {
+                activityLogger.log(Type.KeyReleased, awtEvent.info())
+            }
+            if (awtEvent is KeyEvent) {
+                when (awtEvent.id) {
+                    KeyEvent.KEY_PRESSED -> activityLogger.log(Type.KeyPressed, awtEvent.info())
+                    KeyEvent.KEY_RELEASED -> activityLogger.log(Type.KeyReleased, awtEvent.info())
+                }
+            }
+            false
+        }, trackingDisposable)
+    }
+
+    private fun <L : Any> L.connect(topic: Topic<L>) {
+        val connection = ApplicationManager.getApplication().messageBus.connect()
+        messageBusConnections.add(connection)
+        connection.subscribe(topic, this)
+    }
+
+    private fun KeyEvent.info() = "$id:$keyChar:$keyCode"
+
+    companion object {
+        private val movingActions = MovingActions.values().map { it.name }.toSet()
+        private val typingActions = TypingActions.values().map { it.name }.toSet()
+    }
+
+    private enum class Key(val f: KeyEvent.() -> Boolean) {
+        Meta(KeyEvent::isMetaDown),
+        Control(KeyEvent::isControlDown),
+        Alt(KeyEvent::isAltDown),
+        Shift(KeyEvent::isShiftDown);
+
+        fun check(event: KeyEvent): Key? = if (this@Key.f(event)) this@Key else null
```

Comment:
You don't need to use this@Key, you can use just this, or even omit it like `if (f(event)`